### PR TITLE
[SPARK-40609][SQL] Casts types according to bucket info for Equality expressions

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

It will invalidate the bucket scan if add a cast to bucket column. In fact, sometimes we can avoid adding this cast if it is an equality expression and both sides are integral types.

This PR adds a new type coercion rule(`EqualityTypeCasts`) to support casting types according to bucket info for Equality expressions. For example:
```sql
CREATE TABLE t1(l bigint) USING parquet CLUSTERED BY (l) INTO 200 buckets;
CREATE TABLE t2(dec decimal(18, 0)) USING parquet CLUSTERED BY (dec) INTO 200 buckets;

SET spark.sql.autoBroadcastJoinThreshold=-1;
SELECT * FROM t1 JOIN t2 ON t1.l = t2.dec;
```

Before this PR:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [cast(l#10L as decimal(20,0))], [cast(dec#11 as decimal(20,0))], Inner
   :- Sort [cast(l#10L as decimal(20,0)) ASC NULLS FIRST], false, 0
   :  +- Exchange hashpartitioning(cast(l#10L as decimal(20,0)), 5), ENSURE_REQUIREMENTS, [plan_id=38]
   :     +- Filter isnotnull(l#10L)
   :        +- FileScan parquet spark_catalog.default.t1[l#10L]
   +- Sort [cast(dec#11 as decimal(20,0)) ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(cast(dec#11 as decimal(20,0)), 5), ENSURE_REQUIREMENTS, [plan_id=42]
         +- Filter isnotnull(dec#11)
            +- FileScan parquet spark_catalog.default.t2[dec#11]
```
After this PR:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- SortMergeJoin [l#10L], [try_cast(dec#11 as bigint)], Inner
   :- Sort [l#10L ASC NULLS FIRST], false, 0
   :  +- Filter isnotnull(l#10L)
   :     +- FileScan parquet spark_catalog.default.t1[l#10L]
   +- Sort [try_cast(dec#11 as bigint) ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(try_cast(dec#11 as bigint), 200), ENSURE_REQUIREMENTS, [plan_id=38]
         +- Filter isnotnull(dec#11)
            +- FileScan parquet spark_catalog.default.t2[dec#11]
```

### Why are the changes needed?

Reduce shuffle to improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -751,6 +753,43 @@ abstract class TypeCoercionBase {
...
+      exprs.map(_.dataType).forall {
+        case _: IntegralType => true
+        case DecimalType.Fixed(_, 0) => true
+        case _ => false
+      }
+    }
+
+    override val transform: PartialFunction[Expression, Expression] = {
+      case b @ Equality(left: Attribute, right: Attribute)
+          if b.childrenResolved && left.dataType != right.dataType && isIntegralTypes(b.children) =>
```
Comment (amaliujia) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:771]: Why we don't care what `left.dataType` is?  The test added in `TypeCoercionSuite.scala` only tests Integer, Long and Decimal. If we don't have a constraint on the `left.dataType` probably we need to test more types?

Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:771]: Oh Sorry I mis-read the code here: it check `isIntegralTypes(b.children)` in which `b` is the Equality thus both left and right data types are verified.

**Comment by wangyum** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 2]
```diff
@@ -751,6 +753,49 @@ abstract class TypeCoercionBase {
...
+        case _ => false
+      }
+    }
+
+    override val transform: PartialFunction[Expression, Expression] = {
+      case b @ Equality(l: Attribute, r: Attribute)
+          if b.childrenResolved && l.dataType != r.dataType && isIntegralTypes(b.children) =>
+        // The result type is:
+        // 1. The attribute data type with larger bucket number.
+        // 2. The attribute data type with larger default size if it is not bucketed attribute.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: I don't get it. Can you explain more? Are you saying that adding cast to either side is correct, and we need to make a decision with least cost?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: Yes. adding cast to either side is correct for equality expressions.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: but this is not true under ansi mode, right?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: can we improve the bucket scan planning code to accept safe cast?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: We can use `TRY` mode if under ansi mode:
https://github.com/apache/spark/blob/600dc6cf86bd66d6b92d73cd8dda7ac39a166032/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L785-L791

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: > can we improve the bucket scan planning code to accept safe cast?

Do you mean something like [this changes](https://github.com/apache/spark/pull/32063/files)?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: Ah yes! Can you reopen and update that PR?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: one more pr https://github.com/apache/spark/pull/34755 : )

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: That change is incorrect. For example:
```scala
import org.apache.spark.sql.catalyst.expressions.{Cast, EmptyRow, Literal, Murmur3Hash, Pmod}
import org.apache.spark.sql.types.LongType

println(Pmod(new Murmur3Hash(Seq(Literal(100))), Literal(10)).eval(EmptyRow))
println(Pmod(new Murmur3Hash(Seq(new Cast(Literal(100), LongType))), Literal(10)).eval(EmptyRow))
```

Output:
```
3
0
```

Related code:
https://github.com/apache/spark/blob/69aa727ff495f6698fe9b37e952dfaf36f1dd5eb/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L505-L508

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: good point ! the value is equivalent but the hash code can be different according to the data type

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:774]: New PR: https://github.com/apache/spark/pull/40805

**Pull request closed without merging**